lib/elixir/lib/integer.ex

```diff
@@ -238,67 +238,51 @@ defmodule Integer do
   def parse("", base) when base in 2..36,
     do: :error
 
-  def parse(binary, base) when is_binary(binary) and base in 2..36 do
-    parse_in_base(binary, base)
+  def parse(<<bin::binary>>, base) when base in 2..36 do
+    parse_sign(bin, base)
   end
 
-  def parse(binary, base) when is_binary(binary) do
+  def parse(<<_::binary>>, base) do
     raise ArgumentError, "invalid base #{inspect base}"
   end
 
-  defp parse_in_base("-" <> bin, base) do
-    case do_parse(bin, base) do
-      {number, remainder} -> {-number, remainder}
-      :error -> :error
-    end
+  defp parse_sign(<<?-, rest::binary>>, base) do
+    parse_digits(rest, base, :-)
   end
 
-  defp parse_in_base("+" <> bin, base) do
-    do_parse(bin, base)
+  defp parse_sign(<<?+, rest::binary>>, base) do
+    parse_digits(rest, base, :+)
   end
 
-  defp parse_in_base(binary, base) when is_binary(binary) do
-    do_parse(binary, base)
+  defp parse_sign(<<rest::binary>>, base) do
+    parse_digits(rest, base, :+)
   end
 
-  defp do_parse(<<char, rest::binary>>, base) do
-    if valid_digit_in_base?(char, base) do
-      do_parse(rest, base, parse_digit(char))
-    else
-      :error
+  digits = [{?0..?9, -?0}, {?A..?Z, 10-?A}, {?a..?z, 10-?a}]
```

Comment:
Is there any justification to use variable on the module level? Obviously it's used in the metaprogramming block below, but shouldn't it be a module attribute at least?